src/snowflake/snowpark/_internal/compiler/cte_utils.py

```diff
@@ -71,12 +84,57 @@ def is_duplicate_subtree(encoded_node_id_with_query: str) -> bool:
         return False
 
     traverse(root)
-    duplicated_node = {
+    duplicated_node_ids = {
         encoded_node_id_with_query
         for encoded_node_id_with_query in id_count_map
         if is_duplicate_subtree(encoded_node_id_with_query)
     }
-    return duplicated_node
+
+    if propagate_complexity_hist:
+        return (
+            duplicated_node_ids,
+            get_duplicated_node_complexity_distribution(
+                duplicated_node_ids, id_complexity_map
+            ),
+        )
+    else:
+        return (duplicated_node_ids, None)
+
+
+def get_duplicated_node_complexity_distribution(
+    duplicated_node_id_set: Set[str], id_complexity_map: Dict[str, List[int]]
+) -> List[int]:
+    """
+    Calculate the complexity distribution for the detected repeated node. The complexity are categorized as following:
+    1) low complexity
```

Comment:
just curious, where are these ranges from?

Reply:
the upper bound 10,000,000 is from the default exhaust bound, the rest is really just by sense, if we have other suggestions about the bins, please feel free to put the suggestion, i just don't want to send a big list for the complexity if we have a lot of duplications